Lab6's tools: timistamp, waveform's max, min and rms branches FIXED

marvlad opened the pull request:
## Describe your changes
Tools:
LAPPDClusterTree
LAPPDTraceMax

Waveform amplitude max, min, rms variables added in LAPPDTraceMax.
Branches related to the variables able plus the timestamps added in LAPPDClusterTree

OBS:
Those tools are important for the LAPPD testing at Lab6.

marvlad committed:
Add CalcMinMaxAmp method to LAPPDTraceMax

Added CalcMinMaxAmp method to LAPPDTraceMax class.

marvlad committed:
Implement CalcMinMaxAmp function for waveform analysis

Added a new function to calculate minimum and maximum amplitudes along with RMS, standard deviation, variance, and integral of the waveform.

marvlad committed:
Add noise study variables and metadata timestamps

Added new variables for noise study and metadata timestamps also adding RMS, min and max waveform amplitudes fro dark room data

marvlad committed:
Add new timestamp metadata and related branches

marvlad committed:
Clean up LAPPDClusterTree.cpp formatting

Removed unnecessary closing brace and adjusted formatting.

marvlad committed:
Fix missing newline at end of LAPPDClusterTree.cpp

jminock reviewed (approved):
Looks good!

marc1uk commented on UserTools/LAPPDClusterTree/LAPPDClusterTree.cpp:
```diff
@@ -185,7 +195,19 @@ bool LAPPDClusterTree::Execute()
   m_data->Stores["ANNIEEvent"]->Get("TMR",TMR);
 
+  vector<string> acdcmetadata;
+  m_data->Stores["ANNIEEvent"]->Get("ACDCmetadata",acdcmetadata);
```
check return value

marvlad replied:
I added
bool ok = m_data->Stores["ANNIEEvent"]->Get("ACDCmetadata", acdcmetadata);

if (!ok) {
    std::cerr << "ERROR: Could not retrieve ACDCmetadata from ANNIEEvent store." << std::endl;
    return false;   // or handle error appropriately
}

marc1uk commented on UserTools/LAPPDClusterTree/LAPPDClusterTree.cpp:
```diff
@@ -185,7 +195,19 @@ bool LAPPDClusterTree::Execute()
 
+  //std::cout<<"-------------TIMESTAMP In tree ---------------------------------"<<std::endl;
+  int meta_timestamp_int;
+  std::string meta_timestamp = acdcmetadata[204] + acdcmetadata[206] + acdcmetadata[208] + acdcmetadata[210];
```
use `at` method or check vector size.
Can we have some explanation for magic numbers 204, 206, 208, 210? or define enums instead?

marvlad replied:
I added more comments about the numbers and also I al using `at`

marc1uk commented on UserTools/LAPPDClusterTree/LAPPDClusterTree.h:
```diff
@@ -113,7 +121,8 @@ class LAPPDClusterTree: public Tool {
   Int_t         pulseside_simp[60];
 
-
+  // NEW Mrv!
```
will there be a follow-up pull request to update this once the variable is no longer considered new? :upside_down_face:

marvlad replied:
it is removed the comment

marc1uk commented on UserTools/LAPPDTraceMax/LAPPDTraceMax.cpp:
```diff
@@ -175,12 +192,58 @@ double LAPPDTraceMax::CalcIntegral(Waveform<double> hwav, double lowR, double hi
       tQ+=((hwav.GetSample(i))*Deltat);
     }
-  } else std::cout<<"OUT OF RANGE!!!!"<<std::flush;
+  } else std::cout<<"OUT OF RANGE!!!!";
```
if not endl, then flush at the least is a sensible idea. Alternatively, if this is being called from a tight loop, there should probably be a trailing `std::cout<<std::endl;` or `std::cout<<std::flush` call to ensure the buffers are flushed at the end of the loop, to ensure the output comes out at a controlled point.

marvlad replied:
I don't remember changing this but I added <<std::endl; to all the open std::couts

marc1uk commented on UserTools/LAPPDTraceMax/LAPPDTraceMax.cpp:
```diff
@@ -175,12 +192,58 @@ double LAPPDTraceMax::CalcIntegral(Waveform<double> hwav, double lowR, double hi
+  double mean = 0.0;
+  double variance = 0.0;
+  for (int i=0; i<nbins; i++) mean += theWav->at(i);
+  mean = mean/nbins;
```
`mean` is exactly the same as `sumint` - this doesn't need to be re-calculated.

marvlad replied:
This is cleaned completely. No double calculation and I realize that std::minmax_element firt returns min and second max, i was doing in the wrong order. It is fixed too

marc1uk commented on UserTools/LAPPDTraceMax/LAPPDTraceMax.cpp:
```diff
@@ -175,12 +192,58 @@ double LAPPDTraceMax::CalcIntegral(Waveform<double> hwav, double lowR, double hi
+  double sumInt = 0.0;
+  for (int i=0; i<nbins; i++) sumInt += theWav->at(i);
+
+  double sumint = sumInt / nbins;
```
While technically possible, please try to avoid using variables names that only differ by capitalisation, for clarity.
In this case the temporary `sumInt` is redundant; `sumint` could have been used in its place.

marvlad replied:
Same as before. It is fixed

marc1uk reviewed (commented):
very minor comments

marvlad committed:
Improve ACDCmetadata handling and timestamp extraction

Added error handling for ACDCmetadata retrieval and refactored timestamp extraction from ACDC metadata.

marvlad committed:
Add timestamps_meta to LAPPDClusterTree

marvlad committed:
Add endl to OUT OF RANGE error messages

marvlad committed:
Fix min/max assignments and removing double mean calculation

Refactor CalcMinMaxAmp function to correct min/max assignments and removing double mean calculation.

marvlad committed:
Add meta_timestamp_int variable for timestamp handling

marc1uk commented:
Thanks Marvin :+1:

marvlad commented:
Thank you for checking. I added in hurry those branches to have more sense of the lab6 data. We are still using them and I PR without cleaning
